packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -432,17 +432,42 @@ class NetworkImage extends ImageProvider<NetworkImage> {
     );
   }
 
-  static final http.Client _httpClient = createHttpClient();
+  static final HttpClient _httpClient = new HttpClient();
 
   Future<ui.Codec> _loadAsync(NetworkImage key) async {
     assert(key == this);
 
     final Uri resolved = Uri.base.resolve(key.url);
-    final http.Response response = await _httpClient.get(resolved, headers: headers);
-    if (response == null || response.statusCode != 200)
+    final HttpClientRequest request = await _httpClient.getUrl(resolved);
+    headers?.forEach((String name, String value) {
+      request.headers.add(name, value);
+    });
+    final HttpClientResponse response = await request.close();
+    if (response.statusCode != HttpStatus.OK)
       throw new Exception('HTTP request failed, statusCode: ${response?.statusCode}, $resolved');
 
-    final Uint8List bytes = response.bodyBytes;
+    final Completer<Uint8List> completer = new Completer<Uint8List>.sync();
+    if (response.contentLength == -1) {
+      final ByteConversionSink sink = new ByteConversionSink.withCallback((List<int> chunk) {
+        completer.complete(new Uint8List.fromList(chunk));
+      });
+      response.listen(sink.add, onDone: sink.close, onError: completer.completeError, cancelOnError: true);
+    } else {
+      final Uint8List bytes = new Uint8List(response.contentLength);
+      int offset = 0;
+      response.listen((List<int> chunk) {
+          for (int i = 0; i < chunk.length; i++, offset++) {
+            bytes[offset] = chunk[i];
+          }
+        },
+        onError: completer.completeError,
+        onDone: () {
+          completer.complete(bytes);
+        },
+        cancelOnError: true,
+      );
+    }
+    final Uint8List bytes = await completer.future;
```

Comment:
cc @rmacnak-google who may have a better idea of how to do this without copying bytes multiple times.

Reply:
I think the case where the total length is known is optimal: copy from the OS socket into the chunk, copy from chunks into the linear buffer. The unknown length case has additional copies from growing the buffer and trimming the final result. I believe the chunks are not recycled, so it could be improved by holding onto the chunks and building the linear buffer after receiving the last chunk.

Reply:
Well the "optimal" case wouldn't copy at all, right, we'd just re-use the OS buffer unmodified...